atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -0,0 +1,224 @@
+        return ClusterInfo(cluster, platform, osbs, load)
+
+    def choose_cluster(self, platform):
+        config = get_config(self.workflow)
+        clusters = [self.get_cluster_info(cluster, platform) for cluster in
+                    config.get_enabled_clusters_for_platform(platform)]
+
+        if not clusters:
+            raise RuntimeError('No clusters found for platform {}!'
+                               .format(platform))
+
+        selected = min(clusters, key=lambda c: c.load)
+        self.log.info('platform %s will use cluster %s',
+                      platform, selected.cluster.name)
+        return selected
+
+    def get_release(self):
+        labels = df_parser(self.workflow.builder.df_path, workflow=self.workflow).labels
+        return get_preferred_label(labels, 'release')
+
+    def get_worker_build_kwargs(self, release, platform):
+        build_kwargs = deepcopy(self.build_kwargs)
+
+        build_kwargs.pop('architecture', None)
+
+        build_kwargs['release'] = release
+        build_kwargs['platform'] = platform
+
+        return build_kwargs
+
+    def do_worker_build(self, release, cluster_info):
+        build = None
+        try:
+            kwargs = self.get_worker_build_kwargs(release, cluster_info.platform)
+            build = cluster_info.osbs.create_worker_build(**kwargs)
+        except Exception:
+            self.log.exception('%s - failed to create worker build',
+                               cluster_info.platform)
+
+        build_info = WorkerBuildInfo(build=build, cluster_info=cluster_info)
+        self.worker_builds.append(build_info)
+
+        if build_info.build:
+            try:
+                build_info.wait_to_schedule()
+                build_info.watch_logs(self.log)
+                build_info.wait_to_finish()
+            except Exception:
+                self.log.exception('%s - failed to monitor worker build',
+                                   cluster_info.platform)
+
+    def run(self):
+        release = self.get_release()
+        platforms = self.get_platforms()
+
+        thread_pool = ThreadPool(len(platforms))
+        result = thread_pool.map_async(
+            lambda cluster_info: self.do_worker_build(release, cluster_info),
+            [self.choose_cluster(platform) for platform in platforms]
+        )
+
+        try:
+            while True:
+                # The wait call is a blocking call which prevents signals
+                # from being processed. Wait for short intervals instead
+                # of a single long interval, so build cancellation can
+                # be handled virtually immediately.
+                result.wait(1)
+        # Always clean up worker builds on any error to avoid
+        # runaway worker builds (includes orchestrator build cancellation)
+        except Exception:
+            self.log.info('build cancelled, cancelling worker builds')
+            for build_info in self.worker_builds:
```

Comment:
Since there's a limited amount of time between receiving SIGTERM and SIGKILL, it's probably a good idea to perform the cancellations in parallel, yes.